keps/sig-storage/20200116-node-failover.md

```diff
@@ -0,0 +1,230 @@
+
+[1] [Pod safety design](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/pod-safety.md)
+[2] [Gc orphaned pods](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podgc/gc_controller.go#L97)
+[3] [requirements on node objects](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L167)
+[8] [bootstrap checkpointing](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/0004-bootstrap-checkpointing.md)
+[9] [bootstrap checkpointing deprecation intent](https://github.com/kubernetes/enhancements/issues/1103)
+
+### Goals
+
+* Increase the availability of stateful workloads
+* Automate self-healing for stateful workloads
+
+### Non-Goals
+
+* Introduce a new API to report node state
+* Handle cases where the Kubelet itself is unresponsive
+
+
+## Proposal
+
+user stories:
+
+* as a cluster administrator I want my stateful workload to failover in case of a node is shutdown without any intervention
+
+* as a developer I would like to rely on a self-healing platform for my stateful workloads
+
+This KEP plans to introduce the following capability to the CSI capability to the spec:
+
+```
+type PluginCapability_Service_Type int32
+
+const (
+	// SAFE_DETACH indicates that the volumes targeting
+	// this plugin MAY be safely detached from the node.
+	// If this capability is advertised The CO can assume
+	// that the volume won't be detached if it is still
+	// mounted and accessible from the container
+	PluginCapability_Service_SAFE_DETACH PluginCapability_Service_Type = 3
+)
+```
+
+This should let the external-attacher know if it can assume safe detaches
+
+
+#### kube-controller-manager
+
+
+for kube-controller-manager we need to change the csi volume plugin to make sure that
+its CSI attacher, specifically:
+
+```go
+func (c *csiAttacher) Detach(volumeName string, nodeName types.NodeName) error
+```
+
+is not deleting the `VolumeAttachement` objects related to the PVs by checking If
+the CSI driver has the `SAFE_DETACH` capability
+
+so when calling CPI using `InstanceShutdownByProviderID` If it returns `true` we should apply the existing taint:
+
+```
+taints:
+- key: "cloudprovider.kubernetes.io/node-shutdown"
+  operator: "Exists"
+  effect: "NoSchedule"
+```
+
+After that we can then call:
+
+```go
+if err := c.k8s.StorageV1().VolumeAttachments().Delete(attachID, nil); err != nil {
+  if apierrors.IsNotFound(err) {
+    // object deleted or never existed, done
+    klog.V(4).Info(log("VolumeAttachment object [%v] for volume [%v] not found, object deleted", attachID, volID))
+    return nil
+  }
+  return errors.New(log("detacher.Detach failed to delete VolumeAttachment [%s]: %v", attachID, err))
+}
+```
+
+to avoid leaking taint concepts into the CSI attacher we'll introduce a `DetachCondition` interface
```

Comment:
Will this DetachCondition be defined in apis/storage/types.go?  Please write down the full definition.